From 5f80f092331b26b3f735ff6515ad1eef3a117e53 Mon Sep 17 00:00:00 2001
From: Preeti U Murthy <preetium@andrew.cmu.edu>
Date: Mon, 7 Nov 2016 18:41:04 +0530
Subject: [PATCH 1/2] Offset and number of bytes read correct

---
 file.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/file.c b/file.c
index b4536e9..4a7026b 100644
--- a/file.c
+++ b/file.c
@@ -19,6 +19,7 @@
 #include <linux/string.h>
 #include <linux/slab.h>
 #include <linux/uaccess.h>
+#include <linux/uio.h>
 
 #include <linux/coda.h>
 #include <linux/coda_psdev.h>
@@ -38,8 +39,8 @@ coda_file_read_iter(struct kiocb *iocb, struct iov_iter *to)
 
 	BUG_ON(!cfi || cfi->cfi_magic != CODA_MAGIC);
 
+    printk("The read is at offset %lld, #bytes read = %ld", iocb->ki_pos, iov_iter_count(to));
     size = vfs_iter_read(cfi->cfi_container, to, &iocb->ki_pos);
-    printk("The read is at offset %lld, #bytes read = %ld", iocb->ki_pos, size);
     return size;
 }
 
-- 
2.5.0

